src/main/java/sparksoniq/jsoniq/runtime/iterator/CommaExpressionIterator.java

```diff
@@ -27,49 +28,50 @@
 import java.util.List;
 
 public class CommaExpressionIterator extends LocalRuntimeIterator {
-
     public CommaExpressionIterator(List<RuntimeIterator> childIterators, IteratorMetadata iteratorMetadata) {
         super(childIterators, iteratorMetadata);
     }
 
     @Override
     public Item next() {
-        if(currentChild == null) {
-            currentChild = this._children.get(0);
-            this._children.forEach(c -> c.reset(this._currentDynamicContext));
-            currentChild.open(_currentDynamicContext);
-        }
-        if(currentChild.hasNext())
-            return currentChild.next();
-        else
-        {
-            currentChild.close();
-            if(this._children.indexOf(currentChild) == this._children.size() - 1)
-                throw new IteratorFlowException("Invalid next() call in Comma expression", getMetadata());
-            currentChild = this._children.get(_children.indexOf(currentChild) + 1);
-            currentChild.open(_currentDynamicContext);
-            return currentChild.next();
+        if(_hasNext == true){
+            Item result = _nextResult;  // save the result to be returned
+            setNextResult();            // calculate and store the next result
+            return result;
         }
+        throw new IteratorFlowException("Invalid next() call in Comma expression", getMetadata());
     }
 
     @Override
-    public void open(DynamicContext context){
+    public void open(DynamicContext context) {
         super.open(context);
-        currentChild = null;
-    }
+        this._nextIndex = 0;
 
-    @Override
-    public void reset(DynamicContext context){
-        super.reset(context);
-        currentChild = null;
+        setNextResult();
     }
 
-    @Override
-    public boolean hasNext(){
-        return currentChild == null ||
-                this._children.indexOf(currentChild) < this._children.size() - 1 ||
-                currentChild.hasNext();
+    public void setNextResult() {
+        _nextResult = null;
+
+        for (int i = _nextIndex; i < this._children.size(); i++) {
+            RuntimeIterator childIterator = this._children.get(i);
+            childIterator.open(_currentDynamicContext);
+            if (childIterator.hasNext()) {
+                _nextResult = childIterator.next();
+                _nextIndex = i + 1;
```

Comment:
I am not sure it is quite correct if one of the input expressions returns more than one item. Can you add tests like:

(("foo", "bar"), (), ("foo", "bar", "foobar"), "foo)

?

You need to iterate at two levels:
- within one child iterator
- and across child iterators once a child iterator was fully consumed